hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/service/TestSnapshotDeletingService.java

```diff
@@ -0,0 +1,301 @@
+
+    snapshotDeletingService.setSuccessRunCount(0);
+
+    GenericTestUtils.waitFor(() ->
+            snapshotDeletingService.getSuccessfulRunCount() >= 1,
+        1000, 10000);
+
+    OmSnapshot nextSnapshot = (OmSnapshot) om.getOmSnapshotManager()
+        .checkForSnapshot(VOLUME_NAME, BUCKET_NAME_ONE,
+            getSnapshotPrefix("bucket1snap3"));
+
+    // Check bucket1key1 added to next non deleted snapshot db.
+    RepeatedOmKeyInfo omKeyInfo =
+        nextSnapshot.getMetadataManager()
+            .getDeletedTable().get("/vol1/bucket1/bucket1key1");
+    Assertions.assertNotNull(omKeyInfo);
+
+    // Check bucket1key2 added active db as it can be reclaimed.
+    RepeatedOmKeyInfo omKeyInfo1 = omMetadataManager
+        .getDeletedTable().get("/vol1/bucket1/bucket1key2");
+
+    Assertions.assertNotNull(omKeyInfo1);
+
+  }
+
+  @Test
+  public void testMultipleSnapshotKeyReclaim() throws Exception {
+
+    SnapshotDeletingService snapshotDeletingService = (SnapshotDeletingService)
+        keyManager.getSnapshotDeletingService();
+
+    // Suspending SnapshotDeletingService
+    snapshotDeletingService.suspend();
+    int snapshotCount = createSnapshotDataForBucket1();
+
+    OmKeyArgs bucket2key1 = createVolumeBucketKey(VOLUME_NAME, BUCKET_NAME_TWO,
+        BucketLayout.DEFAULT, "bucket2key1");
+
+    OmKeyArgs bucket2key2 = createKey(VOLUME_NAME, BUCKET_NAME_TWO,
+        "bucket2key2");
+
+    createSnapshot(VOLUME_NAME, BUCKET_NAME_TWO, "bucket2snap1",
+        ++snapshotCount);
+
+    // Both key 1 and key 2 can be reclaimed when Snapshot 1 is deleted.
+    writeClient.deleteKey(bucket2key1);
+    writeClient.deleteKey(bucket2key2);
+
+    createSnapshot(VOLUME_NAME, BUCKET_NAME_TWO, "bucket2snap2",
+        ++snapshotCount);
+
+    String snapshotKey2 = "/vol1/bucket2/bucket2snap1";
+    SnapshotInfo snapshotInfo = om.getMetadataManager()
+        .getSnapshotInfoTable().get(snapshotKey2);
+
+    snapshotInfo
+        .setSnapshotStatus(SnapshotInfo.SnapshotStatus.SNAPSHOT_DELETED);
+    om.getMetadataManager()
+        .getSnapshotInfoTable().put(snapshotKey2, snapshotInfo);
+    snapshotInfo = om.getMetadataManager()
+        .getSnapshotInfoTable().get(snapshotKey2);
+    Assertions.assertEquals(snapshotInfo.getSnapshotStatus(),
+        SnapshotInfo.SnapshotStatus.SNAPSHOT_DELETED);
+
+    snapshotDeletingService.resume();
+
+    snapshotDeletingService.setSuccessRunCount(0L);
+    GenericTestUtils.waitFor(() ->
+            snapshotDeletingService.getSuccessfulRunCount() >= 1,
+        1000, 10000);
+
+    // Check bucket2key1 added active db as it can be reclaimed.
+    RepeatedOmKeyInfo omKeyInfo1 = omMetadataManager
+        .getDeletedTable().get("/vol1/bucket2/bucket2key1");
+
+    // Check bucket2key2 added active db as it can be reclaimed.
+    RepeatedOmKeyInfo omKeyInfo2 = omMetadataManager
+        .getDeletedTable().get("/vol1/bucket2/bucket2key2");
+
+    Assertions.assertNotNull(omKeyInfo1);
```

Comment:
I don't think this test is actually working properly; it continues to pass if you comment out the following lines:

    //snapshotDeletingService.resume();
    //snapshotDeletingService.setSuccessRunCount(0L);
    //GenericTestUtils.waitFor(() ->
    //        snapshotDeletingService.getSuccessfulRunCount() >= 1,
    //    1000, 10000);

I think it is only passing because @smengcl 's pr hasn't been merged yet: https://github.com/apache/ozone/pull/4280  Once that is merged in I believe it will fail, because I don't think it is actually testing the correct thing.  I believe the keys will only be moved to the active fs when bucket2snap2 is deleted.  What do you think?

Reply:
Let me take a look at it, might be missing something in the test.